### ISSUE TITLE
AssertionError (SuperAccessors.scala:380)

### ISSUE THREAD

**Issue by xuwei-k** [scala/scala-dev#348]
https://github.com/xuwei-k/scalikejdbc-assert-error

### Test.scala

```scala
package example

case class Foo(value: Int)

case class MyClass(a: Int)
object MyClass extends scalikejdbc.SQLSyntaxSupport[MyClass]

object Test {
  def bar: scalikejdbc.SQLSyntaxSupport[_] = ???

  def test(): Unit = {
    scalikejdbc.withSQL(
      scalikejdbc.update(bar).set().where.eq(MyClass.column.a, Foo(1))
    )
  }
}
```

### build.sbt

```scala
scalaVersion := "2.12.1"

libraryDependencies += "org.scalikejdbc" %% "scalikejdbc" % "3.0.0-RC3"
```

### error log

```
at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:380)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.$anonfun$transformArgs$1(SuperAccessors.scala:112)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer$$Lambda$988/2116490522.apply(Unknown Source)
	at scala.reflect.internal.TreeInfo.$anonfun$mapMethodParamsAndArgs$1$adapted(TreeInfo.scala:232)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer$$Lambda$989/1885675277.apply(Unknown Source)
	at scala.reflect.internal.TreeInfo.foreachMethodParamAndArg(TreeInfo.scala:246)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transformArgs(SuperAccessors.scala:109)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:381)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:71)
	at scala.reflect.internal.Trees.$anonfun$itransform$2(Trees.scala:1367)
	at scala.reflect.internal.Trees$$Lambda$786/484189179.apply(Unknown Source)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.atOwner(SuperAccessors.scala:414)
	at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
	at scala.reflect.internal.Trees.itransform(Trees.scala:1365)
	at scala.reflect.internal.Trees.itransform$(Trees.scala:1340)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:44)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:389)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:71)
	at scala.reflect.api.Trees$Transformer.$anonfun$transformTrees$1(Trees.scala:2559)
	at scala.reflect.api.Trees$Transformer.transformTrees(Trees.scala:2559)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.$anonfun$transform$4(SuperAccessors.scala:247)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer$$Lambda$985/1732447013.apply(Unknown Source)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.atOwner(SuperAccessors.scala:414)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transformTemplate$1(SuperAccessors.scala:247)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:252)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:71)
	at scala.reflect.api.Trees$Transformer.transformTemplate(Trees.scala:2563)
	at scala.reflect.internal.Trees.$anonfun$itransform$5(Trees.scala:1417)
	at scala.reflect.internal.Trees$$Lambda$807/955593387.apply(Unknown Source)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.atOwner(SuperAccessors.scala:414)
	at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
	at scala.reflect.internal.Trees.itransform(Trees.scala:1416)
	at scala.reflect.internal.Trees.itransform$(Trees.scala:1340)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:44)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:235)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:71)
	at scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2589)
	at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2587)
	at scala.reflect.internal.Trees.$anonfun$itransform$7(Trees.scala:1430)
	at scala.reflect.internal.Trees$$Lambda$806/1837800099.apply(Unknown Source)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.atOwner(SuperAccessors.scala:414)
	at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
	at scala.reflect.internal.Trees.itransform(Trees.scala:1430)
	at scala.reflect.internal.Trees.itransform$(Trees.scala:1340)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.super$transform(TypingTransformers.scala:40)
	at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.$anonfun$transform$2(TypingTransformers.scala:42)
	at scala.tools.nsc.transform.TypingTransformers$TypingTransformer$$Lambda$805/1483009212.apply(Unknown Source)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.atOwner(SuperAccessors.scala:414)
	at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:25)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:389)
	at scala.tools.nsc.typechecker.SuperAccessors$SuperAccTransformer.transform(SuperAccessors.scala:71)
	at scala.tools.nsc.ast.Trees$Transformer.transformUnit(Trees.scala:140)
	at scala.tools.nsc.transform.Transform$Phase.apply(Transform.scala:30)
	at scala.tools.nsc.Global$GlobalPhase.$anonfun$applyPhase$1(Global.scala:416)
	at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:409)
	at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1(Global.scala:380)
	at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1$adapted(Global.scala:380)
	at scala.tools.nsc.Global$GlobalPhase$$Lambda$260/319131353.apply(Unknown Source)
	at scala.collection.Iterator.foreach(Iterator.scala:929)
	at scala.collection.Iterator.foreach$(Iterator.scala:929)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1406)
	at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:380)
	at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1418)
	at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1403)
	at scala.tools.nsc.Global$Run.compileSources(Global.scala:1398)
	at scala.tools.nsc.Global$Run.compile(Global.scala:1492)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:116)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:95)
	at xsbt.CompilerInterface.run(CompilerInterface.scala:26)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:107)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:53)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:47)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply$mcV$sp(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler$$anonfun$compileScala$1$1.apply(MixedAnalyzingCompiler.scala:50)
	at sbt.compiler.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:74)
	at sbt.compiler.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:49)
	at sbt.compiler.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:64)
	at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
	at sbt.compiler.IC$$anonfun$compileInternal$1.apply(IncrementalCompiler.scala:160)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:66)
	at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:64)
	at sbt.inc.IncrementalCommon.cycle(IncrementalCommon.scala:32)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:72)
	at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:71)
	at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:99)
	at sbt.inc.Incremental$.compile(Incremental.scala:71)
	at sbt.inc.IncrementalCompile$.apply(Compile.scala:54)
	at sbt.compiler.IC$.compileInternal(IncrementalCompiler.scala:160)
	at sbt.compiler.IC$.incrementalCompile(IncrementalCompiler.scala:138)
	at sbt.Compiler$.compile(Compiler.scala:155)
	at sbt.Compiler$.compile(Compiler.scala:141)
	at sbt.Defaults$.sbt$Defaults$$compileIncrementalTaskImpl(Defaults.scala:879)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:870)
	at sbt.Defaults$$anonfun$compileIncrementalTask$1.apply(Defaults.scala:868)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[error] (compile:compileIncremental) java.lang.AssertionError: assertion failed: 
```

**Comment by ponkotuy** [comment]
A similar one? https://github.com/ponkotuy/ScalaCompilerBug

```scala
import scalikejdbc._
object Main extends App {
  println("a")
  val u = Model.defaultAlias
  Model.findAllBy(sqls.eq(u.id, u))
}
```

```
> compile
[info] Compiling 1 Scala source to /home/yosuke/sbt/target/scala-2.12/classes...
[trace] Stack trace suppressed: run last compile:compileIncremental for the full output.
[error] (compile:compileIncremental) java.lang.AssertionError: assertion failed:
[error]   transformCaseApply: name = u tree = u / class scala.reflect.internal.Trees$Ident
[error]      while compiling: /home/yosuke/sbt/src/main/scala/Main.scala
[error]         during phase: refchecks
[error]      library version: version 2.12.1
[error]     compiler version: version 2.12.1
[error]   reconstructed args: -bootclasspath /usr/lib/jvm/java-8-openjdk/jre/lib/resources.jar:/usr/lib/jvm/java-8-openjdk/jre/lib/rt.jar:/usr/lib/jvm/java-8-openjdk/jre/lib/sunrsasign.jar:/usr/lib/jvm/java-8-openjdk/jre/lib/jsse.jar:/usr/lib/jvm/java-8-openjdk/jre/lib/jce.jar:/usr/lib/jvm/java-8-openjdk/jre/lib/charsets.jar:/usr/lib/jvm/java-8-openjdk/jre/lib/jfr.jar:/usr/lib/jvm/java-8-openjdk/jre/classes:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.12.1/scala-library-2.12.1.jar -classpath /home/yosuke/sbt/target/scala-2.12/classes:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/joda-time/joda-time/2.9.9/joda-time-2.9.9.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.12.1/scala-reflect-2.12.1.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scalikejdbc/scalikejdbc-core_2.12/2.5.1/scalikejdbc-core_2.12-2.5.1.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/skinny-framework/skinny-orm_2.12/2.3.6/skinny-orm_2.12-2.3.6.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/slf4j/slf4j-api/1.7.25/slf4j-api-1.7.25.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/skinny-framework/skinny-common_2.12/2.3.6/skinny-common_2.12-2.3.6.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scalikejdbc/scalikejdbc-interpolation_2.12/2.5.1/scalikejdbc-interpolation_2.12-2.5.1.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scalikejdbc/scalikejdbc_2.12/2.5.1/scalikejdbc_2.12-2.5.1.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/apache/commons/commons-dbcp2/2.1.1/commons-dbcp2-2.1.1.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/skinny-framework/skinny-micro-common_2.12/1.2.5/skinny-micro-common_2.12-1.2.5.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/commons-logging/commons-logging/1.2/commons-logging-1.2.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/joda/joda-convert/1.8.1/joda-convert-1.8.1.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/com/typesafe/config/1.3.1/config-1.3.1.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scalikejdbc/scalikejdbc-interpolation-macro_2.12/2.5.1/scalikejdbc-interpolation-macro_2.12-2.5.1.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scalikejdbc/scalikejdbc-config_2.12/2.5.1/scalikejdbc-config_2.12-2.5.1.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/flywaydb/flyway-core/4.0.3/flyway-core-4.0.3.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/apache/commons/commons-pool2/2.4.2/commons-pool2-2.4.2.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-parser-combinators_2.12/1.0.5/scala-parser-combinators_2.12-1.0.5.jar:/home/yosuke/.coursier/cache/v1/https/repo1.maven.org/maven2/org/scalikejdbc/scalikejdbc-syntax-support-macro_2.12/2.5.1/scalikejdbc-syntax-support-macro_2.12-2.5.1.jar
[error]
[error]   last tree to typer: This(object Main)
[error]        tree position: line 6 of /home/yosuke/sbt/src/main/scala/Main.scala
[error]             tree tpe: Main.type
[error]               symbol: object Main
[error]    symbol definition: class Main extends App (a ModuleClassSymbol)
[error]       symbol package: <empty>
[error]        symbol owners: object Main
[error]            call site: object Main in package <empty>
[error]
[error] == Source file context for tree position ==
[error]
[error]      3 object Main extends App {
[error]      4   println("a")
[error]      5   val u = Model.defaultAlias
[error]      6   Model.findAllBy(sqls.eq(u.id, u))
[error]      7 }
[error] Total time: 0 s, completed 2017/04/04 2:26:58
```

**Comment by diesalbla** [comment]
@xuwei-k This was already fixed in Scala 2.12.6 and Scala 2.12.3, or earlier. I have not been able to test it with the 2.13 milestones, because the `scalikejdbc` dependency is not available for it yet.  

 I have cloned your repo with the error project, modified the `build.sbt` to set the Scala version to `2.12.6`, and run `sbt:compile`. There is no longer a compiler crash, but the following error instead: 
```
[info] Done updating.
[info] Compiling 1 Scala source to /Users/diesalbla/floss/scalikejdbc-assert-error/target/scala-2.12/classes...
[error] /Users/diesalbla/floss/scalikejdbc-assert-error/Test.scala:13:
[error] --------------------------------------------------------
[error]  Implicit ParameterBinderFactory[A] for the parameter type A is missing.
[error]  You need to define ParameterBinderFactory for the type or use AsIsParameterBinder.
[error]
[error]   (example1)
[error]     implicit val intParameterBinderFactory: ParameterBinderFactory[Int] = ParameterBinderFactory {
[error]        value => (stmt, idx) => stmt.setInt(idx, value)
[error]      }
[error]
[error]   (example2)
[error]     case class Price(value: Int)
[error]     object Price {
[error]       implicit val converter: Binders[Price] = Binders.int.xmap(Price.apply, _.value)
[error]     }
[error]
[error]   (example3)
[error]     val value: Any = 123
[error]     val key: SQLSyntax = sqls"column_name"
[error]     key -> AsIsParameterBinder(value)
[error] --------------------------------------------------------
[error]       scalikejdbc.update(bar).set().where.eq(MyClass.column.a, Foo(1))
[error]                                             ^
[error] one error found
```

**Comment by SethTisue** [comment]
thanks @diesalbla for checking on it
